From 4a8529bbbdc5019fbc8eadc451993e5411c5549f Mon Sep 17 00:00:00 2001
From: Daniel Stenberg <daniel@haxx.se>
Date: Mon, 26 Aug 2019 16:00:05 +0200
Subject: [PATCH] http2: when marked for closure and wanted to close == OK

It could otherwise return an error even when closed correctly if GOAWAY
had been received previously.

Reported-by: Tom van der Woerdt
Fixes #4267
---
 lib/http2.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/curl-src/lib/http2.c b/curl-src/lib/http2.c
index 930e851653..31d2d698ad 100644
--- a/curl-src/lib/http2.c
+++ b/curl-src/lib/http2.c
@@ -1559,6 +1559,11 @@ static ssize_t http2_recv(struct connectdata *conn, int sockindex,
   if(should_close_session(httpc)) {
     H2BUGF(infof(data,
                  "http2_recv: nothing to do in this session\n"));
+    if(conn->bits.close) {
+      /* already marked for closure, return OK and we're done */
+      *err = CURLE_OK;
+      return 0;
+    }
     *err = CURLE_HTTP2;
     return -1;
   }
